### ISSUE TITLE
prod() is broken on ROCM when used on tensors of int type on gfx908

### ISSUE THREAD

**Issue by IMbackK** [pytorch/pytorch#120794]
### 🐛 Describe the bug

ROCM attempts to run a GFX906 kernel on any device (tested on GFX900, GFX908 and GFX1030) when prod is used on any int datatype

Repoduction:

```
AMD_SERIALIZE_KERNEL=3 AMD_LOG_LEVEL=8 python
Python 3.11.7 (main, Jan 29 2024, 16:03:57) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.ones((2), dtype=torch.int32).to("cuda:0").prod()
Error: The program ISA amdgcn-amd-amdhsa--gfx906:sramecc+:xnack- is not compatible with the device ISA amdgcn-amd-amdhsa--gfx908:sramecc+:xnack-
RuntimeError: CUDA driver error: 303
```

Running the example specifically on a gfx906 device dose work, as dose casting to any float datatype

### Versions

```
Collecting environment information...
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.0.32830-

OS: Arch Linux (x86_64)
GCC version: (GCC) 13.2.1 20230801
Clang version: 16.0.6
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.7 (main, Jan 29 2024, 16:03:57) [GCC 13.2.1 20230801] (64-bit runtime)
Python platform: Linux-6.6.0-arch1-1-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI100 (gfx908:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.0.32830
MIOpen runtime version: 3.0.0
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             48
On-line CPU(s) list:                0-47
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7402P 24-Core Processor
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU(s) scaling MHz:                 58%
CPU max MHz:                        2800.0000
CPU min MHz:                        1500.0000
BogoMIPS:                           5602.33
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          768 KiB (24 instances)
L1i cache:                          768 KiB (24 instances)
L2 cache:                           12 MiB (24 instances)
L3 cache:                           128 MiB (8 instances)
NUMA node(s):                       8
NUMA node0 CPU(s):                  0-2,24-26
NUMA node1 CPU(s):                  3-5,27-29
NUMA node2 CPU(s):                  6-8,30-32
NUMA node3 CPU(s):                  9-11,33-35
NUMA node4 CPU(s):                  12-14,36-38
NUMA node5 CPU(s):                  15-17,39-41
NUMA node6 CPU(s):                  18-20,42-44
NUMA node7 CPU(s):                  21-23,45-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Vulnerable
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.2.0
[conda] Could not collect
```

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

**Comment by jeffdaily** [comment]
Fun.  Trying to create an environment now that matches yours.  Unless you have a handy docker image that can reproduce your issue?

**Comment by jeffdaily** [comment]
What is your OS?  How did you install rocm and pytorch?

**Comment by IMbackK** [comment]
im on arch linux. Rocm 6.0 is supplied by the distro packages. torch 2.2.0 is self compiled with some minor patches, however i can also reproduce this problem with the official torch-2.3.0.dev20240228 and torch-2.3.0.dev20240207 wheel.

I also noticed that this problem dosent exist on a system with only one gpu type, on a different system with just a gfx1030 it seams to work. So i will amend my reproduction steps to include:

1. have a system with gfx1030 (gpu 0) and at least one gfx908 gpu
2. mask away the gfx1030 gpu with ROCR_VISIBLE_DEVICES
3. run the steps above.

**Comment by hongxiayang** [comment]
how about using HIP_VISIBLE_DEVICES environment variable?

**Comment by IMbackK** [comment]
I can confirm that HIP_VISIBLE_DEVICES behaves the same way. Also i have removed the gfx906 device from the setup and still observe the same result, its presence is thus not required to trigger this bug.

It is my suspicion that the presence of a gfx908 device specifically is required to trigger this bug.

**Comment by IMbackK** [comment]
I did yes, i compiled for gfx1030, gfx900, gfx906 and gfx908. also all other pytorch functions work fine on the 908 as far as i can tell its only prod with int.

It also fails with the official wheel which is compiled for all the regular targets ofc.

**Comment by hongxiayang** [comment]
run with AMD_LOG_LEVEL should tell more about kernel launch, and that may shed light where to break in gdb.

**Comment by IMbackK** [comment]
```
:3:rocdevice.cpp            :445 : 10006729215 us: [pid:17440 tid:0x7f0039a19740] Initializing HSA stack.
:3:rocdevice.cpp            :211 : 10006757361 us: [pid:17440 tid:0x7f0039a19740] Numa selects cpu agent[1]=0x55e2964f5e40(fine=0x55e298ce0bf0,coarse=0x55e297b681a0) for gpu agent=0x55e298efcb10 CPU<->GPU XGMI=0
:3:rocdevice.cpp            :1715: 10006760402 us: [pid:17440 tid:0x7f0039a19740] Gfx Major/Minor/Stepping: 9/0/8
:3:rocdevice.cpp            :1717: 10006760409 us: [pid:17440 tid:0x7f0039a19740] HMM support: 1, XNACK: 0, Direct host access: 0
:3:rocdevice.cpp            :1719: 10006760412 us: [pid:17440 tid:0x7f0039a19740] Max SDMA Read Mask: 0xff, Max SDMA Write Mask: 0xff
:4:rocdevice.cpp            :2099: 10006761502 us: [pid:17440 tid:0x7f0039a19740] Allocate hsa host memory 0x7eff862be000, size 0x38
:4:rocdevice.cpp            :2099: 10006762044 us: [pid:17440 tid:0x7f0039a19740] Allocate hsa host memory 0x7efca2e00000, size 0x101000
:4:rocdevice.cpp            :2099: 10006762409 us: [pid:17440 tid:0x7f0039a19740] Allocate hsa host memory 0x7efca2c00000, size 0x101000
:4:runtime.cpp              :83  : 10006762496 us: [pid:17440 tid:0x7f0039a19740] init
:3:hip_context.cpp          :48  : 10006762501 us: [pid:17440 tid:0x7f0039a19740] Direct Dispatch: 1
:3:hip_device_runtime.cpp   :637 : 10008009771 us: [pid:17440 tid:0x7f0039a19740]  hipGetDeviceCount ( 0x7fffa8c4c1e0 ) 
:3:hip_device_runtime.cpp   :639 : 10008009810 us: [pid:17440 tid:0x7f0039a19740] hipGetDeviceCount: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :637 : 10008009875 us: [pid:17440 tid:0x7f0039a19740]  hipGetDeviceCount ( 0x7eff59442994 ) 
:3:hip_device_runtime.cpp   :639 : 10008009881 us: [pid:17440 tid:0x7f0039a19740] hipGetDeviceCount: Returned hipSuccess : 
:3:hip_device.cpp           :463 : 10008009889 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevicePropertiesR0600 ( 0x7fffa8c4bd08, 0 ) 
:3:hip_device.cpp           :465 : 10008009896 us: [pid:17440 tid:0x7f0039a19740] hipGetDevicePropertiesR0600: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :637 : 10008010073 us: [pid:17440 tid:0x7f0039a19740]  hipGetDeviceCount ( 0x7fffa8c4c218 ) 
:3:hip_device_runtime.cpp   :639 : 10008010078 us: [pid:17440 tid:0x7f0039a19740] hipGetDeviceCount: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008010091 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4bfe4 ) 
:3:hip_device_runtime.cpp   :630 : 10008010093 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :637 : 10008010098 us: [pid:17440 tid:0x7f0039a19740]  hipGetDeviceCount ( 0x7fffa8c4bd34 ) 
:3:hip_device_runtime.cpp   :639 : 10008010100 us: [pid:17440 tid:0x7f0039a19740] hipGetDeviceCount: Returned hipSuccess : 
:3:hip_context.cpp          :344 : 10008011075 us: [pid:17440 tid:0x7f0039a19740]  hipDevicePrimaryCtxGetState ( 0, 0x7fffa8c4bdf8, 0x7fffa8c4bdfc ) 
:3:hip_context.cpp          :358 : 10008011083 us: [pid:17440 tid:0x7f0039a19740] hipDevicePrimaryCtxGetState: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008011086 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4c004 ) 
:3:hip_device_runtime.cpp   :630 : 10008011088 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_context.cpp          :344 : 10008011091 us: [pid:17440 tid:0x7f0039a19740]  hipDevicePrimaryCtxGetState ( 0, 0x7fffa8c4be18, 0x7fffa8c4be1c ) 
:3:hip_context.cpp          :358 : 10008011093 us: [pid:17440 tid:0x7f0039a19740] hipDevicePrimaryCtxGetState: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008011098 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4bfa4 ) 
:3:hip_device_runtime.cpp   :630 : 10008011100 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_context.cpp          :344 : 10008011103 us: [pid:17440 tid:0x7f0039a19740]  hipDevicePrimaryCtxGetState ( 0, 0x7fffa8c4bdb8, 0x7fffa8c4bdbc ) 
:3:hip_context.cpp          :358 : 10008011105 us: [pid:17440 tid:0x7f0039a19740] hipDevicePrimaryCtxGetState: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008015909 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b624 ) 
:3:hip_device_runtime.cpp   :630 : 10008015916 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008015921 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b504 ) 
:3:hip_device_runtime.cpp   :630 : 10008015923 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008015932 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b4f4 ) 
:3:hip_device_runtime.cpp   :630 : 10008015933 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_stream.cpp           :395 : 10008015944 us: [pid:17440 tid:0x7f0039a19740]  hipDeviceGetStreamPriorityRange ( 0x7fffa8c4b298, 0x7fffa8c4b29c ) 
:3:hip_stream.cpp           :403 : 10008015951 us: [pid:17440 tid:0x7f0039a19740] hipDeviceGetStreamPriorityRange: Returned hipSuccess : 
:3:hip_error.cpp            :35  : 10008015974 us: [pid:17440 tid:0x7f0039a19740]  hipGetLastError (  ) 
:3:hip_device_runtime.cpp   :622 : 10008015977 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4aca4 ) 
:3:hip_device_runtime.cpp   :630 : 10008015979 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_graph.cpp            :808 : 10008016303 us: [pid:17440 tid:0x7f0039a19740]  hipStreamIsCapturing ( stream:<null>, 0x7fffa8c4ace8 ) 
:3:hip_graph.cpp            :809 : 10008016310 us: [pid:17440 tid:0x7f0039a19740] hipStreamIsCapturing: Returned hipSuccess : 
:3:hip_memory.cpp           :586 : 10008016638 us: [pid:17440 tid:0x7f0039a19740]  hipMalloc ( 0x7fffa8c4adf8, 2097152 ) 
:4:rocdevice.cpp            :2227: 10008016763 us: [pid:17440 tid:0x7f0039a19740] Allocate hsa device memory 0x7efca2800000, size 0x200000
:3:rocdevice.cpp            :2266: 10008016768 us: [pid:17440 tid:0x7f0039a19740] device=0x55e298f741a0, freeMem_ = 0x7fee00000
:3:hip_memory.cpp           :588 : 10008016779 us: [pid:17440 tid:0x7f0039a19740] hipMalloc: Returned hipSuccess : 0x7efca2800000: duration: 141 us
:3:hip_device_runtime.cpp   :652 : 10008016800 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008016804 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :652 : 10008016806 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008016808 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008018199 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b5a4 ) 
:3:hip_device_runtime.cpp   :630 : 10008018204 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008018207 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b5a4 ) 
:3:hip_device_runtime.cpp   :630 : 10008018209 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_memory.cpp           :672 : 10008018531 us: [pid:17440 tid:0x7f0039a19740]  hipMemcpyWithStream ( 0x7efca2800000, 0x55e298cdb640, 8, hipMemcpyHostToDevice, stream:<null> ) 
:3:rocdevice.cpp            :2768: 10008018541 us: [pid:17440 tid:0x7f0039a19740] number of allocated hardware queues with low priority: 0, with normal priority: 0, with high priority: 0, maximum per priority is: 4
:3:rocdevice.cpp            :2846: 10008034159 us: [pid:17440 tid:0x7f0039a19740] created hardware queue 0x7efe825d3000 with size 16384 with priority 1, cooperative: 0
:3:rocdevice.cpp            :2938: 10008034170 us: [pid:17440 tid:0x7f0039a19740] acquireQueue refCount: 0x7efe825d3000 (1)
:4:rocdevice.cpp            :2099: 10008034437 us: [pid:17440 tid:0x7f0039a19740] Allocate hsa host memory 0x7efca2400000, size 0x100000
:3:devprogram.cpp           :2686: 10008533933 us: [pid:17440 tid:0x7f0039a19740] Using Code Object V5.
:4:command.cpp              :347 : 10008536488 us: [pid:17440 tid:0x7f0039a19740] Command (CopyHostToDevice) enqueued: 0x55e29942b5e0
:4:rocblit.cpp              :799 : 10008537096 us: [pid:17440 tid:0x7f0039a19740] HSA Async Copy staged H2D dst=0x7efca2800000, src=0x7efca2e00000, size=8, completion_signal=0x7efca41ff780
:4:rocvirtual.cpp           :556 : 10008537106 us: [pid:17440 tid:0x7f0039a19740] Host wait on completion_signal=0x7efca41ff780
:3:rocvirtual.hpp           :66  : 10008537108 us: [pid:17440 tid:0x7f0039a19740] Host active wait for Signal = (0x7efca41ff780) for -1 ns
:4:command.cpp              :285 : 10008537128 us: [pid:17440 tid:0x7f0039a19740] Queue marker to command queue: 0x55e298cda170
:4:command.cpp              :347 : 10008537130 us: [pid:17440 tid:0x7f0039a19740] Command (InternalMarker) enqueued: 0x55e2997f4f60
:4:command.cpp              :175 : 10008537137 us: [pid:17440 tid:0x7f0039a19740] Command 0x55e29942b5e0 complete
:4:command.cpp              :169 : 10008537141 us: [pid:17440 tid:0x7f0039a19740] Command 0x55e2997f4f60 complete (Wall: 10008537140, CPU: 0, GPU: 0 us)
:4:command.cpp              :249 : 10008537145 us: [pid:17440 tid:0x7f0039a19740] Waiting for event 0x55e29942b5e0 to complete, current status 0
:4:command.cpp              :264 : 10008537152 us: [pid:17440 tid:0x7f0039a19740] Event 0x55e29942b5e0 wait completed
:3:hip_memory.cpp           :683 : 10008537158 us: [pid:17440 tid:0x7f0039a19740] hipMemcpyWithStream: Returned hipSuccess : : duration: 518627 us
:3:hip_device_runtime.cpp   :652 : 10008537169 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008537174 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008539414 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4c164 ) 
:3:hip_device_runtime.cpp   :630 : 10008539419 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008539435 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4bd44 ) 
:3:hip_device_runtime.cpp   :630 : 10008539437 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008539441 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4bc44 ) 
:3:hip_device_runtime.cpp   :630 : 10008539443 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008539448 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4ba94 ) 
:3:hip_device_runtime.cpp   :630 : 10008539451 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :652 : 10008539465 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008539469 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :652 : 10008539472 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008539475 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008540567 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b0a4 ) 
:3:hip_device_runtime.cpp   :630 : 10008540572 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008540575 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4af84 ) 
:3:hip_device_runtime.cpp   :630 : 10008540576 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008540579 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4af74 ) 
:3:hip_device_runtime.cpp   :630 : 10008540581 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :652 : 10008540587 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008540589 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :652 : 10008540591 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008540592 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008540642 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b074 ) 
:3:hip_device_runtime.cpp   :630 : 10008540649 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008540654 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4ab44 ) 
:3:hip_device_runtime.cpp   :630 : 10008540657 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_platform.cpp         :193 : 10008540669 us: [pid:17440 tid:0x7f0039a19740]  __hipPushCallConfiguration ( {1,1,1}, {512,1,1}, 0, stream:<null> ) 
:3:hip_platform.cpp         :197 : 10008540675 us: [pid:17440 tid:0x7f0039a19740] __hipPushCallConfiguration: Returned hipSuccess : 
:3:hip_platform.cpp         :202 : 10008540683 us: [pid:17440 tid:0x7f0039a19740]  __hipPopCallConfiguration ( {1418605999,2,2831461304}, {2831461344,32767,3224297866}, 0x7fffa8c4aba0, 0x7fffa8c4ab88 ) 
:3:hip_platform.cpp         :211 : 10008540688 us: [pid:17440 tid:0x7f0039a19740] __hipPopCallConfiguration: Returned hipSuccess : 
:3:hip_module.cpp           :668 : 10008541293 us: [pid:17440 tid:0x7f0039a19740]  hipLaunchKernel ( 0x7f00229c8ea8, {1,1,1}, {512,1,1}, 0x7fffa8c4aea0, 0, stream:<null> ) 
:3:devprogram.cpp           :2686: 10008543795 us: [pid:17440 tid:0x7f0039a19740] Using Code Object V5.
:4:command.cpp              :347 : 10008546804 us: [pid:17440 tid:0x7f0039a19740] Command (KernelExecution) enqueued: 0x55e299d46170
:3:rocvirtual.cpp           :784 : 10008546815 us: [pid:17440 tid:0x7f0039a19740] Arg0:  N = val:2
:3:rocvirtual.cpp           :784 : 10008546818 us: [pid:17440 tid:0x7f0039a19740] Arg1:   = val:
:3:rocvirtual.cpp           :2925: 10008546821 us: [pid:17440 tid:0x7f0039a19740] ShaderName : _ZN2at6native6legacy18elementwise_kernelILi512ELi1EZNS0_15gpu_kernel_implIZZZNS0_23direct_copy_kernel_cudaERNS_18TensorIteratorBaseEENKUlvE0_clEvENKUlvE2_clEvEUllE_EEvS5_RKT_EUliE_EEviT1_
:4:rocvirtual.cpp           :870 : 10008546832 us: [pid:17440 tid:0x7f0039a19740] HWq=0x7efca2600000, Dispatch Header = 0x1502 (type=2, barrier=1, acquire=2, release=2), setup=3, grid=[512, 1, 1], workgroup=[512, 1, 1], private_seg_size=0, group_seg_size=0, kernel_obj=0x7efca2f60000, kernarg_address=0x7efca2400000, completion_signal=0x0
:4:command.cpp              :285 : 10008546838 us: [pid:17440 tid:0x7f0039a19740] Queue marker to command queue: 0x55e298cda170
:4:command.cpp              :347 : 10008546842 us: [pid:17440 tid:0x7f0039a19740] Command (InternalMarker) enqueued: 0x55e2997f4f60
:4:rocvirtual.cpp           :1026: 10008546848 us: [pid:17440 tid:0x7f0039a19740] HWq=0x7efca2600000, BarrierAND Header = 0x1503 (type=3, barrier=1, acquire=2, release=2), dep_signal=[0x0, 0x0, 0x0, 0x0, 0x0], completion_signal=0x7efca41ff700
:4:rocvirtual.cpp           :556 : 10008546851 us: [pid:17440 tid:0x7f0039a19740] Host wait on completion_signal=0x7efca41ff700
:3:rocvirtual.hpp           :66  : 10008546853 us: [pid:17440 tid:0x7f0039a19740] Host active wait for Signal = (0x7efca41ff700) for -1 ns
:4:command.cpp              :175 : 10008546862 us: [pid:17440 tid:0x7f0039a19740] Command 0x55e299d46170 complete
:4:command.cpp              :169 : 10008546865 us: [pid:17440 tid:0x7f0039a19740] Command 0x55e2997f4f60 complete (Wall: 10008546864, CPU: 0, GPU: 0 us)
:4:command.cpp              :249 : 10008546867 us: [pid:17440 tid:0x7f0039a19740] Waiting for event 0x55e299d46170 to complete, current status 0
:4:command.cpp              :264 : 10008546870 us: [pid:17440 tid:0x7f0039a19740] Event 0x55e299d46170 wait completed
:3:hip_module.cpp           :669 : 10008546872 us: [pid:17440 tid:0x7f0039a19740] hipLaunchKernel: Returned hipSuccess : 
:3:hip_error.cpp            :35  : 10008546875 us: [pid:17440 tid:0x7f0039a19740]  hipGetLastError (  ) 
:3:hip_error.cpp            :35  : 10008546878 us: [pid:17440 tid:0x7f0039a19740]  hipGetLastError (  ) 
:3:hip_device_runtime.cpp   :652 : 10008546882 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008546886 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008546916 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b844 ) 
:3:hip_device_runtime.cpp   :630 : 10008546919 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device.cpp           :463 : 10008546929 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevicePropertiesR0600 ( 0x7fffa8c4b210, 0 ) 
:3:hip_device.cpp           :465 : 10008546934 us: [pid:17440 tid:0x7f0039a19740] hipGetDevicePropertiesR0600: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008546938 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b854 ) 
:3:hip_device_runtime.cpp   :630 : 10008546941 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008546944 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b854 ) 
:3:hip_device_runtime.cpp   :630 : 10008546947 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008546966 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b7f4 ) 
:3:hip_device_runtime.cpp   :630 : 10008546969 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hip_context.cpp          :186 : 10008547414 us: [pid:17440 tid:0x7f0039a19740]  hipCtxGetCurrent ( 0x7fffa8c4b020 ) 
:3:hip_context.cpp          :190 : 10008547424 us: [pid:17440 tid:0x7f0039a19740] hipCtxGetCurrent: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :622 : 10008547427 us: [pid:17440 tid:0x7f0039a19740]  hipGetDevice ( 0x7fffa8c4b074 ) 
:3:hip_device_runtime.cpp   :630 : 10008547430 us: [pid:17440 tid:0x7f0039a19740] hipGetDevice: Returned hipSuccess : 
:3:hiprtc.cpp               :230 : 10008547710 us: [pid:17440 tid:0x7f0039a19740] hiprtcVersion ( 0x7fffa8c4b118, 0x7fffa8c4b11c )
:3:hiprtc.cpp               :240 : 10008547741 us: [pid:17440 tid:0x7f0039a19740] hiprtcVersion: Returned HIPRTC_SUCCESS
:3:hip_module.cpp           :57  : 10008548229 us: [pid:17440 tid:0x7f0039a19740]  hipModuleLoadData ( 0x7fffa8c4b190, 0x55e2996eb770 ) 
:1:devprogram.cpp           :1876: 10008548344 us: [pid:17440 tid:0x7f0039a19740] Error: The program ISA amdgcn-amd-amdhsa--gfx906:sramecc+:xnack- is not compatible with the device ISA amdgcn-amd-amdhsa--gfx908:sramecc+:xnack-
Error: create kernel metadata map using COMgr
Error: Cannot Find Global Var Sizes
Error: Cannot create kernels.

:3:hip_module.cpp           :58  : 10008548356 us: [pid:17440 tid:0x7f0039a19740] hipModuleLoadData: Returned hipErrorSharedObjectInitFailed : 
:3:hip_device_runtime.cpp   :652 : 10008564508 us: [pid:17440 tid:0x7f0039a19740]  hipSetDevice ( 0 ) 
:3:hip_device_runtime.cpp   :656 : 10008564515 us: [pid:17440 tid:0x7f0039a19740] hipSetDevice: Returned hipSuccess : 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: CUDA driver error: 303

```
i gues 
"_ZN2at6native6legacy18elementwise_kernelILi512ELi1EZNS0_15gpu_kernel_implIZZZNS0_23direct_copy_kernel_cudaERNS_18TensorIteratorBaseEENKUlvE0_clEvENKUlvE2_clEvEUllE_EEvS5_RKT_EUliE_EEviT1_" is the suspect shader, however i have no idea how to decode where this comes from (torch? miopen? blas?)

**Comment by hongxiayang** [comment]
I quickly tested on gfx908(MI100) with the sample code you have with the  `rocm/pytorch:latest` docker, I can not reproduce your problem:
```
root@xxx: python
Python 3.9.18 (main, Sep 11 2023, 13:41:44) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.ones((2), dtype=torch.int32).to("cuda:0").prod()
tensor(1, device='cuda:0')
```
Comparing with your environment, I can see some differences. I saw your OS is `Arch Linux (x86_64)` while in my docker container, mine is `Ubuntu 20.04.6 LTS (x86_64)`. Would you mind trying to verify using the `rocm/pytorch:latest` docker container?

**Comment by IMbackK** [comment]
So i tried rocm/pytorch:latest and rocm/pytorch:rocm6.0.2_ubuntu22.04_py3.10_pytorch_2.1.2 and in both cases it did indeed work fine. See docker.log. So i did some further investigation:

1. in the non-working (arch) case i get the error "Error: create kernel metadata map using COMgr"
    * in the working case (docker) i have "Loading COMGR library."
    * in the non-working case we never see this
2. on arch linux if i run the example as root with "sudo AMD_SERIALIZE_KERNEL=3 AMD_LOG_LEVEL=8 ROCR_VISIBLE_DEVICES=1 python" it works just like in the docker image
3. running pytorch under a newly created user with minimal env is also broken

It thus seams like some permissions issue with COMGR it to blame for this bug

[docker.log](https://github.com/pytorch/pytorch/files/14498452/docker.log)
[archlinux.log](https://github.com/pytorch/pytorch/files/14498454/archlinux.log)

**Comment by hongxiayang** [comment]
@IMbackK Very nice investigations.

**Comment by jeffdaily** [comment]
Does the rocminfo utility work correctly or also report permission issues on /dev/kfd or /dev/dri?

**Comment by IMbackK** [comment]
sure rocminfo works fine, as do all other operation in torch other than prod on int tensors. Other hip applications other than torch also work fine

**Comment by IMbackK** [comment]
@hongxiayang @jeffdaily  Further investigation has revealed the reason for the behvior:

1. Torch creates a cache of compiled kernels at $HOME/.cache/torch/kernels
2. When prod is executed for the first time the file reduction_prod_arch9.0_nvrtc9.0_ptx_29483_42ed31f661ab2ead9946fe5434715b3b8c5cc5a0 is created
3. crucially the name of this file dose not differ regardless of if prod() is executed on gfx906 or gfx908 or (or presumably also gfx900) for the first time
4. when prod() is then executed again this code object will be selected from the cache, regardless of if the current device is gfx906 or gfx908 (or gfx900)
5. prod() thus remains broken forever for all other gfx9 devices if a gfx906 was the first user of prod() regardless of if its still present in the system
6. Executing as root changed $HOME, avoiding this issue

I can now reproduce the problem in docker by:

1. executing prod on gfx906
2. restart pytorch
3. executing prod on gfx908

Notes:

1. It seams that this cache mechanism is barely used (maybe only used by prod()), allowing this issue to go unnoticed for a long time. Despite me using torch extensively for various training tasks my $HOME/.cache/torch/kernels contains only prod() kernels (for various data types)
2. I have attached the offending kernel below, but it can also be easily recreated by following the reproduction steps.

[kernel.tar.gz](https://github.com/pytorch/pytorch/files/14502647/kernel.tar.gz)

**Comment by IMbackK** [comment]
@hongxiayang @jeffdaily 

The kernel is selected here based on this constructed filename:
https://github.com/pytorch/pytorch/blob/cac36e232ec047517eec9f4e2cf6f63762264629/aten/src/ATen/native/cuda/jit_utils.cpp#L1503

`cuda_major << "." << cuda_minor; ` is simply insufficient information to determine if the cached kernel is viable or not.

**Comment by IMbackK** [comment]
pagging @mruberry as this is your code

**Comment by IMbackK** [comment]
I would like to propose the following solution (PR later):

```
From 81c45e7a3cb3f1eae6f652a949920e94418796b5 Mon Sep 17 00:00:00 2001
From: uvos <devnull@uvos.xyz>
Date: Wed, 6 Mar 2024 01:28:11 +0100
Subject: [PATCH] [ATen][Native][CUDA] Fix kernel cache selecting kernels for
 incorect rocm architectures

fixes #120794

Signed-off-by: Carl Philipp Klemm <carl@uvos.xyz>
---
 aten/src/ATen/native/cuda/jit_utils.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/aten/src/ATen/native/cuda/jit_utils.cpp b/aten/src/ATen/native/cuda/jit_utils.cpp
index 61781e03b4a..fdf975eeb5c 100644
--- a/aten/src/ATen/native/cuda/jit_utils.cpp
+++ b/aten/src/ATen/native/cuda/jit_utils.cpp
@@ -1500,7 +1500,11 @@ NvrtcFunction jit_pwise_function(
     std::stringstream ss;
     ss << *cache_dir << "/";
     ss << kernel_name;
+#ifdef USE_ROCM
+    ss << "_arch" << prop->gcnArchName;
+#else
     ss << "_arch" << cuda_major << "." << cuda_minor;
+#endif
     ss << "_nvrtc" << nvrtc_major << "." << nvrtc_minor;
     ss << (compile_to_sass ? "_sass" : "_ptx");
     ss << "_" << code.length();
-- 
2.43.2
```

**Comment by hongxiayang** [comment]
@IMbackK Very thorough. Great work!

**Comment by jeffdaily** [comment]
I think this is related to the AT_JITERATOR code paths that some kernels take.  To reduce binary size, some kernels are JIT'd.

**Comment by jeffdaily** [comment]
@IMbackK I would approve your PR if you submit one.  Tag me in it if it isn't automatic.  I can approve CI jobs to start running but can't approve PRs for merging.
